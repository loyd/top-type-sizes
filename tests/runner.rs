use top_type_sizes::*;

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            static SAMPLE: &'static str =
                include_str!(concat!("samples/", stringify!($name), ".txt"));

            let refined_sample = reader::read(SAMPLE.as_bytes()).unwrap();
            let types = parser::parse(&refined_sample).unwrap();
            insta::assert_debug_snapshot!(format!("{}_types", stringify!($name)), types);

            // TODO: check the transformer.

            let output = formatter::format(types, &Default::default());
            println!("{output}");
            // TODO: make snapshots of output.
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
