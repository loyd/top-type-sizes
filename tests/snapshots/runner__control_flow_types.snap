---
source: tests/runner.rs
expression: types
---
[
    Type {
        name: "std::ops::ControlFlow<std::option::Option<std::convert::Infallible>, alloc::collections::btree::navigate::LazyLeafHandle<alloc::collections::btree::node::marker::Dying, std::ffi::OsString, std::option::Option<std::ffi::OsString>>>",
        size: 32,
        align: 8,
        kind: Enum(
            EnumType {
                discriminant_size: None,
                variants: [
                    EnumVariant {
                        name: "Continue",
                        size: 32,
                        items: [
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "0",
                                    size: 32,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                    EnumVariant {
                        name: "Break",
                        size: 0,
                        items: [
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "0",
                                    size: 0,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                ],
            },
        ),
        end_padding: None,
    },
]
